vite.plugins.ts

```diff
@@ -79,3 +81,63 @@ export const minifyHTML = (): {
     return minify(html, { collapseWhitespace: true });
   },
 });
+
+/**
+ * Forwards requests to the local replica.
+ * Denies access to raw URLs.
+ *
+ * @param replicaOrigin Replica URL to forward requests to
+ * @param forwardRules List of rules (i.e. hostname to canisterId mappings)
+ *                     to forward requests to a specific canister
+ */
+export const replicaForwardPlugin = ({
+  replicaOrigin,
+  forwardRules,
+}: {
+  replicaOrigin: string;
+  forwardRules: Array<{ canisterId: string; hosts: string[] }>;
+}) => ({
+  name: "replica-forward",
+  configureServer(server: ViteDevServer) {
+    const proxy = httpProxy.createProxyServer({
+      secure: false,
+    });
+
+    server.middlewares.use((req, res, next) => {
+      if (
+        /* deny requests to raw URLs */
+        req.headers["host"]?.endsWith(".raw.ic0.app") ||
+        req.headers["host"]?.endsWith(".raw.icp0.io")
+      ) {
+        console.log(
+          `Denying access to raw URL ${req.method} https://${req.headers.host}${req.url}`
+        );
+        res.statusCode = 400;
+        res.end("Raw IC URLs are not supported");
+        return;
+      }
+
+      const matchingRule = forwardRules.find((rule) => {
+        let host = req.headers["host"];
+        return isNullish(host) || rule.hosts.includes(host);
```

Comment:
if I understand this correctly, when the `host` header is not set, you return the first rule in the list. Is this correct?

How about moving the host reading out of the callback? Also would it make sense to return `next()` if the `host` is nullish?

Reply:
Ah yes, this is a bug. Fixed, thanks!